colexecbase: speed up identity cast with selection vector

yuzefovich opened the pull request:
Previously, the identity cast operator could be quite inefficient when
the input batch had a selection vector. In particular, this was the case
because we copied all values in `[0, sel[n-1]]` range, including all the
not selected values. This stemmed from the fact that we don't want to
perform the "deselection" on the batch, so we must copy the selected
values into the same positions in the output as they have in the input.
As it turns out we already have a primitive that can be used for this
while only copying the selected values, so this commit updates the
identity cast operator to use that primitive when necessary.

Fixes: #98210.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98513)
<!-- Reviewable:end -->

msirek reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98513#-:-NQRBGfE0wGUjA63mVQf:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

mgartner commented on pkg/sql/colexec/colexecbase/cast_test.go:
```diff
@@ -48,7 +48,8 @@ func TestRandomizedCast(t *testing.T) {
 			if _, ok := cast.LookupCastVolatility(from, to); ok {
-				if colexecbase.IsCastSupported(from, to) {
+				// Skip identical casts since they are not interesting.
+				if colexecbase.IsCastSupported(from, to) && !colexecbase.IsIdentityCast(from, to) {
```
I'm worried about losing this testing coverage. Do you have an explanation for why these casts are so slow? I'd expect them to be faster than any other casts.

yuzefovich reviewed (commented):
bors r-

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98513#-:-NQRcv2U3UQtw_PsmV61:b-ydi9aw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/colexec/colexecbase/cast_test.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98513#-NQRAQCrG9yD3LqNT5QB-r1-52:-NQRaveU43M5tpekQxPi:b-mj18lv) ([raw file](https://github.com/cockroachdb/cockroach/blob/707e0447190c34dd0e4096702406d3d456f8b172/pkg/sql/colexec/colexecbase/cast_test.go#L52)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm worried about losing this testing coverage. Do you have an explanation for why these casts are so slow? I'd expect them to be faster than any other casts.
</blockquote></details>

The commit message explains why they are slow - we copy over a lot of garbage. This is especially problematic in tests given that our `RunTests` test harness 1) runs all tests with `coldata.BatchSize()` set to `1` (which never happens in production) 2) fully randomizes the selection vector (so we always get something like `sel = [255, 1023]` - which seems possible in production although unlikely (?)).

The current way the identity cast works when the batch has selection vector is as follows:
- say we have `input := []int64{x, x, x, ..., 0, x, x, x, ..., 1}, sel := [511, 1023]`
- we need to update `output[511] = 0` and `output[1023] = 1`
- we don't really have a good primitive for this, so before this PR we would just do `copy(output[:1024], input[:1024])`, after this PR we will do `copy(output[511:1024], input[511:1024])`. Note that this still involves copying a lot of garbage, but less on average.
- one primitive we have is `coldata.SetValueAt` and `coldata.GetValueAt` which would allow us to copy only the selected value, but I quickly tested it and it was still slow (at least in tests).
- the proper fix for this would be to code-generate type-specific loops.

Now that I typed all of this out, I'll just do it.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexecbase: speed up identity cast with selection vector

Previously, the identity cast operator could be quite inefficient when
the input batch had a selection vector. In particular, this was the case
because we copied all values in `[0, sel[n-1]]` range, including all the
not selected values. This stemmed from the fact that we don't want to
perform the "deselection" on the batch, so we must copy the selected
values into the same positions in the output as they have in the input.
As it turns out we already have a primitive that can be used for this
while only copying the selected values, so this commit updates the
identity cast operator to use that primitive when necessary.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98513)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[`pkg/sql/colexec/colexecbase/cast_test.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98513#-NQRAQCrG9yD3LqNT5QB-r1-52:-NQRgJ881OuozPEg0Z7p:b-ibsag8) ([raw file](https://github.com/cockroachdb/cockroach/blob/707e0447190c34dd0e4096702406d3d456f8b172/pkg/sql/colexec/colexecbase/cast_test.go#L52)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The commit message explains why they are slow - we copy over a lot of garbage. This is especially problematic in tests given that our `RunTests` test harness 1) runs all tests with `coldata.BatchSize()` set to `1` (which never happens in production) 2) fully randomizes the selection vector (so we always get something like `sel = [255, 1023]` - which seems possible in production although unlikely (?)).

The current way the identity cast works when the batch has selection vector is as follows:
- say we have `input := []int64{x, x, x, ..., 0, x, x, x, ..., 1}, sel := [511, 1023]`
- we need to update `output[511] = 0` and `output[1023] = 1`
- we don't really have a good primitive for this, so before this PR we would just do `copy(output[:1024], input[:1024])`, after this PR we will do `copy(output[511:1024], input[511:1024])`. Note that this still involves copying a lot of garbage, but less on average.
- one primitive we have is `coldata.SetValueAt` and `coldata.GetValueAt` which would allow us to copy only the selected value, but I quickly tested it and it was still slow (at least in tests).
- the proper fix for this would be to code-generate type-specific loops.

Now that I typed all of this out, I'll just do it.
</blockquote></details>

As it turns out we actually have a primitive for this, so I updated the PR.

Also, when typing up the updated commit message I realized that my initial "improved" version of this PR had a bug (namely, then we would copy the output values starting at dest index 0 rather than at `sel[0]`), so thanks Marcus for pushing on this.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98513#-:-NQVMoDc-mqCCP_uFYx9:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[`pkg/sql/colexec/colexecbase/cast_test.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98513#-NQRAQCrG9yD3LqNT5QB-r1-52:-NQVMLVf34YCwDV3thUj:b-rapr5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/707e0447190c34dd0e4096702406d3d456f8b172/pkg/sql/colexec/colexecbase/cast_test.go#L52)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

As it turns out we actually have a primitive for this, so I updated the PR.

Also, when typing up the updated commit message I realized that my initial "improved" version of this PR had a bug (namely, then we would copy the output values starting at dest index 0 rather than at `sel[0]`), so thanks Marcus for pushing on this.
</blockquote></details>

Glad you found a better solution!

I wonder if we should revisit the testing batch size of `1`. It really kills the performance on many tests. Maybe we can set it up to run with this batchsize only in CI or in a roachtest?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98513#-:-NQVnOIY64Q7TluokJ1t:b-tlx7ko)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[`pkg/sql/colexec/colexecbase/cast_test.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98513#-NQRAQCrG9yD3LqNT5QB-r1-52:-NQVoxDJ6OLw6-r-HRgq:bhak70r) ([raw file](https://github.com/cockroachdb/cockroach/blob/707e0447190c34dd0e4096702406d3d456f8b172/pkg/sql/colexec/colexecbase/cast_test.go#L52)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Glad you found a better solution!

I wonder if we should revisit the testing batch size of `1`. It really kills the performance on many tests. Maybe we can set it up to run with this batchsize only in CI or in a roachtest?
</blockquote></details>

Thanks again for questioning this. I have to retract another of my earlier claims. I just profiled the slow (before this PR) test execution, and it doesn't collaborate "batch size of 1" slowness:

![Screenshot 2023-03-14 at 8.55.26 AM.png](https://files.reviewableusercontent.io/images/reviewable/5135385/7jEMXVjfnxrGEQAcoT72ZN-W/Screenshot%202023-03-14%20at%208.55.26%20AM.png)

We spent most of the time determining the size of `DTSVector` - this can be easily optimized and I'll do so in a separate PR.

Re: batch size of 1 is quite beneficial for testing, so I think it's worth the overhead to use it all the time. Instead, we should be more diligent than I was initially with every slowdown / timeout and investigate to the root cause.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98513)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[`pkg/sql/colexec/colexecbase/cast_test.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98513#-NQRAQCrG9yD3LqNT5QB-r1-52:-NQW5-u85UYJuCMsKYLA:b-v0r7zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/707e0447190c34dd0e4096702406d3d456f8b172/pkg/sql/colexec/colexecbase/cast_test.go#L52)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Thanks again for questioning this. I have to retract another of my earlier claims. I just profiled the slow (before this PR) test execution, and it doesn't collaborate "batch size of 1" slowness:

![Screenshot 2023-03-14 at 8.55.26 AM.png](https://files.reviewableusercontent.io/images/reviewable/5135385/7jEMXVjfnxrGEQAcoT72ZN-W/Screenshot%202023-03-14%20at%208.55.26%20AM.png)

We spent most of the time determining the size of `DTSVector` - this can be easily optimized and I'll do so in a separate PR.

Re: batch size of 1 is quite beneficial for testing, so I think it's worth the overhead to use it all the time. Instead, we should be more diligent than I was initially with every slowdown / timeout and investigate to the root cause.
</blockquote></details>

It might not be that simple to speed up this after all.

The true root cause for the slowdown is being fixed by this PR. Namely, this `DTSVector.Size` shows up so prominently because we are copying a lot of garbage values which leads to us recomputing the size of effectively the whole vector whereas only a single value should be changed. With this PR in-place we no longer update the vector of length `coldata.BatchSize` when the batch has only length 1. I filed #98595 to track speeding up `DTSVector.Size` though since it can still affect the performance in production (although my guess is that it won't be as pronounced as it was here).


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98513)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[`pkg/sql/colexec/colexecbase/cast_test.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98513#-NQRAQCrG9yD3LqNT5QB-r1-52:-NQWy5ie-UdGs7aaUH0b:b-hmvmii) ([raw file](https://github.com/cockroachdb/cockroach/blob/707e0447190c34dd0e4096702406d3d456f8b172/pkg/sql/colexec/colexecbase/cast_test.go#L52)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

It might not be that simple to speed up this after all.

The true root cause for the slowdown is being fixed by this PR. Namely, this `DTSVector.Size` shows up so prominently because we are copying a lot of garbage values which leads to us recomputing the size of effectively the whole vector whereas only a single value should be changed. With this PR in-place we no longer update the vector of length `coldata.BatchSize` when the batch has only length 1. I filed #98595 to track speeding up `DTSVector.Size` though since it can still affect the performance in production (although my guess is that it won't be as pronounced as it was here).
</blockquote></details>

Alright, I opened up #98609 to speed up `DTSVector.Size`. With #98609 but without the proper fix in this PR the test time went down from 90s+ to 22s.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r+
